### DISCUSSION TITLE
Setting value of an AuxVariable in an external program

### DISCUSSION THREAD

**Discussion by bielsnohr** [idaholab/moose#17795] [Q&A General] [answered]
I am investigating how my MOOSE app would call an external program by dealing with a simple toy problem. My external program uses libMesh and only has one function that takes a libMesh `System`, the name of a variable in that system, and then the time step.

```cpp
void
sweeping_gaussian(System & sys, const std::string var_name, Real time)
```

This function sets the value of the named variable on each node using the node coordinates.

My objective is to use this function to set the value of an `AuxVariable` within a sub-app that I will then `MultiAppCopyTransfer` up to the main app. My sub-app defines a `FastIonProblem` class that inherits from `ExternalProblem`. It takes the name of the `AuxVariable` as a `Param`, and then retrieves the associated underlying libMesh `System` in the initialiser list. Naively, I then pass that `System` directly in my call to the external program function above. Given the function directly "understands" the native MOOSE format of the `System`, I don't think there needs to be a separate in-memory representation of the "ExternalMesh" like is most other ExternalProblems. Please correct me if wrong on that point.

```cpp
#include "FastIonProblem.h"
#include "sweeping_gaussian.h"

registerMooseObject("ExternalFastIonApp", FastIonProblem);

InputParameters
FastIonProblem::validParams()
{
  InputParameters params = ExternalProblem::validParams();

  params.addClassDescription("Retrieve fast ion heat fluxes from an external program");
  params.addRequiredParam<VariableName>(
      "sync_variable", "The variable that the external solution will be synced to");
  return params;
}

FastIonProblem::FastIonProblem(const InputParameters & parameters)
  : ExternalProblem(parameters),
    _sync_to_var_name(getParam<VariableName>("sync_variable")),
    _problem_system(getSystem(_sync_to_var_name))
{
}


void
FastIonProblem::externalSolve()
{
  sweeping_gaussian(_problem_system, _sync_to_var_name, time());
}
```

So far so good, everything compiles, but upon trying to run the MultiApp with this input for the subapp:

```
[Mesh]
  type = GeneratedMesh
  dim = 3
  nx = 10
  ny = 10
  nz = 10
[]

[Problem]
  type = FastIonProblem
  sync_variable = fi_heat_flux
[]

[AuxVariables]
  [fi_heat_flux]
  []
[]

[Executioner]
  type = Transient
  num_steps = 50
  solve_type = NEWTON
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]
```

I get the following error:

```
ext_fast_ion0: *** ERROR ***
ext_fast_ion0: The following error occurred in the object "MOOSE Problem", of type "FastIonProblem".
ext_fast_ion0: 
ext_fast_ion0: Unable to find a system containing the variable fi_heat_flux
```

Evidently, the `AuxVariable` isn't present on the `System` I am querying and passing to the external program. A few questions related to this:

- Am I going about this in the right way?
- If yes, how do I get the system that the AuxVariable is on? I suspect related to the `addExternalVariable()` method that can be overridden for `ExternalProblem` children, but not quite sure how to go about doing that.

Thanks in advance!

---

*UPDATE 2021-05-10 17:38 BST*

I have made some progress myself by realising that I probably needed to call `getAuxiliarySystem()` instead of `getSystem()`. So, my constructor now looks like:

```cpp
FastIonProblem::FastIonProblem(const InputParameters & parameters)
  : ExternalProblem(parameters),
    _sync_to_var_name(getParam<VariableName>("sync_variable")),
    _problem_system(getAuxiliarySystem().system())
{
}
```

This gets rid of my previous error, but now I am getting this error:

```
[0]PETSC ERROR: Object is in wrong state
[0]PETSC ERROR: Not for unassembled vector
...
[0]PETSC ERROR: #1 VecCopy() line 1557 in /opt/civet/build_1/conda_builds/conda_envs/next-a7feace123/conda-bld/moose-petsc_1611687302001/work/src/vec/vec/interface/vector.c
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
```

I am wondering if this has to do with how I am setting the solution in my external program. This is what the external function looks like:

```cpp
void
sweeping_gaussian(System & sys, const std::string var_name, Real time)
{
  // Get a constant reference to the mesh object, system number, and variable
  // number
  const MeshBase & mesh = sys.get_mesh();
  const unsigned int & system_number = sys.number();
  const unsigned short int & variable_number = sys.variable_number(var_name);

  // Boundary information and the left side id
  BoundaryInfo boundary_info = mesh.get_boundary_info();
  boundary_id_type left_side = boundary_info.get_id_by_name("left");

  // Temporary variables for loop calculations
  dof_id_type dof_i;
  Real gauss;

  // Loop through the local nodes
  for (const auto & node_ptr : mesh.local_node_ptr_range())
  {
    const Node & node = *node_ptr;

    // Check if the node is on the left side
    if (boundary_info.has_boundary_id(node_ptr, left_side))
    {
      // Retrieve the dof_id for this node
      dof_i = node.dof_number(system_number, variable_number, 0);

      // Set solution vector at this index to some combination of the
      // coordinate values
      gauss = 10000.0 * exp(-(pow((node(1) - time / 50) / 0.1, 2)));
      sys.solution->set(dof_i, gauss);
    }
  }
};
```

**Comment by aeslaughter** [comment]
@roystgnr Can you help with this?

**Comment by roystgnr** [comment, marked as answer]
Do you simply need to close the vector you're setting before you leave your program?  With PETSc (or any other distributed numerics), most write operations are merely initiated by the setter function, not necessarily completed (because you don't want every set() to potentially delay waiting on a network reply).

You might try doing a `sys.solution->close()` (which completes all the set()s) and a `sys.update()` (which syncs changes from solution to `current_local_solution`) at the end of `sweeping_gaussian()`

**Reply by bielsnohr** [reply]
Yes, not closing the vector in `sweeping_gaussian()` was indeed the problem. Putting 
```cpp
sys.solution->close()
sys.update()
```
 at the end of the function resolved the error and now my MultiApp runs. Thank you!
